File: packages/backend-node/src/domains/ephemeris.ts
import type {
  EphemerisApi,
  SpiceIntCell,
  SpiceStateVector,
  SpiceVector3,
  SpiceWindow,
  SpkPackedDescriptor,
  SpkUnpackedDescriptor,
  SpkezrResult,
  SpkposResult,
} from "@rybosome/tspice-backend-contract";
import { invariant } from "@rybosome/tspice-core";

import type { NativeAddon } from "../runtime/addon.js";

export function createEphemerisApi(native: NativeAddon): EphemerisApi {
  return {
    spkezr: (target, et, ref, abcorr, observer) => {
      const out = native.spkezr(target, et, ref, abcorr, observer);
      invariant(out && typeof out === "object", "Expected spkezr() to return an object");
      invariant(Array.isArray(out.state) && out.state.length === 6, "Expected spkezr().state to be a length-6 array");
      invariant(typeof out.lt === "number", "Expected spkezr().lt to be a number");
      const state = out.state as SpiceStateVector;
      const result: SpkezrResult = { state, lt: out.lt };
      return result;
    },

    spkpos: (target, et, ref, abcorr, observer) => {
      const out = native.spkpos(target, et, ref, abcorr, observer);
      invariant(out && typeof out === "object", "Expected spkpos() to return an object");
      invariant(Array.isArray(out.pos) && out.pos.length === 3, "Expected spkpos().pos to be a length-3 array");
      invariant(typeof out.lt === "number", "Expected spkpos().lt to be a number");
      const pos = out.pos as SpiceVector3;
      const result: SpkposResult = { pos, lt: out.lt };
      return result;
    },

    spkez: (target, et, ref, abcorr, observer) => {
      const out = native.spkez(target, et, ref, abcorr, observer);
      invariant(out && typeof out === "object", "Expected spkez() to return an object");
      invariant(Array.isArray(out.state) && out.state.length === 6, "Expected spkez().state to be a length-6 array");
      invariant(typeof out.lt === "number", "Expected spkez().lt to be a number");
      const state = out.state as SpiceStateVector;
      const result: SpkezrResult = { state, lt: out.lt };
      return result;
    },

    spkezp: (target, et, ref, abcorr, observer) => {
      const out = native.spkezp(target, et, ref, abcorr, observer);
      invariant(out && typeof out === "object", "Expected spkezp() to return an object");
      invariant(Array.isArray(out.pos) && out.pos.length === 3, "Expected spkezp().pos to be a length-3 array");
      invariant(typeof out.lt === "number", "Expected spkezp().lt to be a number");
      const pos = out.pos as SpiceVector3;
      const result: SpkposResult = { pos, lt: out.lt };
      return result;
    },

    spkgeo: (target, et, ref, observer) => {
      const out = native.spkgeo(target, et, ref, observer);
      invariant(out && typeof out === "object", "Expected spkgeo() to return an object");
      invariant(Array.isArray(out.state) && out.state.length === 6, "Expected spkgeo().state to be a length-6 array");
      invariant(typeof out.lt === "number", "Expected spkgeo().lt to be a number");
      const state = out.state as SpiceStateVector;
      const result: SpkezrResult = { state, lt: out.lt };
      return result;
    },

    spkgps: (target, et, ref, observer) => {
      const out = native.spkgps(target, et, ref, observer);
      invariant(out && typeof out === "object", "Expected spkgps() to return an object");
      invariant(Array.isArray(out.pos) && out.pos.length === 3, "Expected spkgps().pos to be a length-3 array");
      invariant(typeof out.lt === "number", "Expected spkgps().lt to be a number");
      const pos = out.pos as SpiceVector3;
      const result: SpkposResult = { pos, lt: out.lt };
      return result;
    },

    spkssb: (target, et, ref) => {
      const out = native.spkssb(target, et, ref);
      invariant(Array.isArray(out) && out.length === 6, "Expected spkssb() to return a length-6 array");
      return out as SpiceStateVector;
    },

    spkcov: (spk: string, idcode: number, cover: SpiceWindow) => {
      native.spkcov(spk, idcode, cover as unknown as number);
    },

    spkobj: (spk: string, ids: SpiceIntCell) => {
      native.spkobj(spk, ids as unknown as number);
    },

    spksfs: (body: number, et: number) => {
      const out = native.spksfs(body, et);
      invariant(out && typeof out === "object", "Expected spksfs() to return an object");
      invariant(typeof out.found === "boolean", "Expected spksfs().found to be a boolean");
      if (!out.found) {
        return { found: false };
      }

      invariant(typeof out.handle === "number", "Expected spksfs().handle to be a number when found");
      invariant(Array.isArray(out.descr) && out.descr.length === 5, "Expected spksfs().descr to be a length-5 array when found");
      invariant(typeof out.ident === "string", "Expected spksfs().ident to be a string when found");

      return {
        found: true,
        handle: out.handle,
        descr: out.descr as unknown as SpkPackedDescriptor,
        ident: out.ident,
      };
    },

    spkpds: (body: number, center: number, frame: string, type: number, first: number, last: number) => {
      const out = native.spkpds(body, center, frame, type, first, last);
      invariant(Array.isArray(out) && out.length === 5, "Expected spkpds() to return a length-5 array");
      return out as unknown as SpkPackedDescriptor;
    },

    spkuds: (descr: SpkPackedDescriptor) => {
      const out = native.spkuds(descr as unknown as number[]);
      invariant(out && typeof out === "object", "Expected spkuds() to return an object");
      const obj = out as Record<string, unknown>;

      const keys = ["body", "center", "frame", "type", "first", "last", "baddr", "eaddr"] as const;
      for (const k of keys) {
        invariant(typeof obj[k] === "number", `Expected spkuds().${k} to be a number`);
      }

      return out as SpkUnpackedDescriptor;
    },
  };
}

File: packages/py-parity-checking/sidecars/spiceypy/src/py_parity_sidecar/handlers/ephemeris.py
from __future__ import annotations

from pathlib import Path
from typing import Any

import spiceypy as sp
from spiceypy.utils.exceptions import NotFoundError
from spiceypy.utils.support_types import SPICEDOUBLE_CELL, SPICEINT_CELL

from ..models import (
    StepEphemerisSpkcls,
    StepEphemerisSpkcov,
    StepEphemerisSpkez,
    StepEphemerisSpkezp,
    StepEphemerisSpkezr,
    StepEphemerisSpkgeo,
    StepEphemerisSpkgps,
    StepEphemerisSpkobj,
    StepEphemerisSpkopa,
    StepEphemerisSpkopn,
    StepEphemerisSpkpds,
    StepEphemerisSpkpos,
    StepEphemerisSpksfs,
    StepEphemerisSpkssb,
    StepEphemerisSpkuds,
    StepEphemerisSpkw08,
    StepOutput,
    WorkflowStep,
)
from ..runtime import SidecarRuntimeContext, resolve_path_ref


def _to_float(value: Any, *, label: str) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise TypeError(f"{label} must be numeric")
    return float(value)


def _to_float_list(value: Any, *, expected_len: int, label: str) -> list[float]:
    if not hasattr(value, "__iter__"):
        raise TypeError(f"{label} must be iterable")

    out = [_to_float(item, label=f"{label}[{idx}]") for idx, item in enumerate(value)]
    if len(out) != expected_len:
        raise ValueError(f"{label} must have length {expected_len}, got {len(out)}")
    return out


def _normalize_state_lt(value: Any, *, label: str) -> tuple[list[float], float]:
    if not isinstance(value, tuple) or len(value) != 2:
        raise ValueError(f"Unexpected {label} return shape: {value!r}")
    state_raw, lt_raw = value
    state = _to_float_list(state_raw, expected_len=6, label=f"{label}.state")
    lt = _to_float(lt_raw, label=f"{label}.lt")
    return state, lt


def _normalize_pos_lt(value: Any, *, label: str) -> tuple[list[float], float]:
    if not isinstance(value, tuple) or len(value) != 2:
        raise ValueError(f"Unexpected {label} return shape: {value!r}")
    pos_raw, lt_raw = value
    pos = _to_float_list(pos_raw, expected_len=3, label=f"{label}.pos")
    lt = _to_float(lt_raw, label=f"{label}.lt")
    return pos, lt


def _get_window(context: SidecarRuntimeContext, window_id: str, *, max_intervals: int) -> Any:
    windows = context.state.cellsWindows.windows
    existing = windows.get(window_id)
    if existing is not None:
        return existing

    if max_intervals <= 0:
        raise ValueError("maxIntervals must be > 0")

    window = SPICEDOUBLE_CELL(max_intervals * 2)
    windows[window_id] = window
    context.register_finalizer(
        f"cellsWindows.release:{window_id}",
        lambda: context.state.cellsWindows.windows.pop(window_id, None),
    )
    return window


def _get_int_cell(context: SidecarRuntimeContext, cell_id: str, *, max_cardinality: int) -> Any:
    cells = context.state.ephemeris.intCells
    existing = cells.get(cell_id)
    if existing is not None:
        return existing

    if max_cardinality <= 0:
        raise ValueError("maxCardinality must be > 0")

    cell = SPICEINT_CELL(max_cardinality)
    cells[cell_id] = cell
    context.register_finalizer(
        f"ephemeris.releaseIntCell:{cell_id}",
        lambda: context.state.ephemeris.intCells.pop(cell_id, None),
    )
    return cell


def _require_spk_handle(context: SidecarRuntimeContext, handle_id: str) -> int:
    handle = context.state.ephemeris.spkHandles.get(handle_id)
    if handle is None:
        raise ValueError(f"SPK handle does not exist: {handle_id}")
    return handle


def _resolve_path(context: SidecarRuntimeContext, path_ref_input: Any) -> str:
    return resolve_path_ref(context.paths, path_ref_input)


def _close_spk_handle_best_effort(handle: int) -> None:
    try:
        sp.spkcls(handle)
    except BaseException:
        # best-effort cleanup only
        pass


def _set_spk_handle(context: SidecarRuntimeContext, handle_id: str, handle: int) -> None:
    existing = context.state.ephemeris.spkHandles.get(handle_id)
    if existing is not None and existing != handle:
        _close_spk_handle_best_effort(existing)

    context.state.ephemeris.spkHandles[handle_id] = handle

    def _finalize() -> None:
        current = context.state.ephemeris.spkHandles.get(handle_id)
        if current is None or current != handle:
            return
        _close_spk_handle_best_effort(handle)
        context.state.ephemeris.spkHandles.pop(handle_id, None)

    context.register_finalizer(f"ephemeris.spkcls:{handle_id}:{handle}", _finalize)


def _normalize_spksfs_result(value: Any) -> dict[str, Any]:
    if isinstance(value, tuple):
        if len(value) == 4:
            _handle, descr, ident, found_raw = value
            found = bool(found_raw)
            if not found:
                return {"found": False}
            return {
                "found": True,
                "descr": _to_float_list(descr, expected_len=5, label="spksfs.descr"),
                "ident": str(ident),
            }

        if len(value) == 3:
            _handle, descr, ident = value
            return {
                "found": True,
                "descr": _to_float_list(descr, expected_len=5, label="spksfs.descr"),
                "ident": str(ident),
            }

    raise ValueError(f"Unexpected spksfs return shape: {value!r}")


def run_ephemeris_step(step: WorkflowStep, context: SidecarRuntimeContext) -> StepOutput | None:
    if isinstance(step, StepEphemerisSpkcls):
        handle = _require_spk_handle(context, step.handleId)
        sp.spkcls(handle)
        context.state.ephemeris.spkHandles.pop(step.handleId, None)
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepEphemerisSpkcov):
        cover = _get_window(
            context,
            step.coverWindowId,
            max_intervals=step.maxIntervals if step.maxIntervals is not None else 16,
        )
        sp.spkcov(_resolve_path(context, step.spk), step.idcode, cover)
        interval_count = int(sp.wncard(cover))
        intervals: list[list[float]] = []
        for idx in range(interval_count):
            left, right = sp.wnfetd(cover, idx)
            intervals.append([float(left), float(right)])
        return StepOutput(op=step.op, value={"intervals": intervals})

    if isinstance(step, StepEphemerisSpkez):
        state, lt = _normalize_state_lt(
            sp.spkez(step.target, step.et, step.ref, step.abcorr, step.observer),
            label="spkez",
        )
        context.state.ephemeris.requestedTargets.add(step.target)
        return StepOutput(op=step.op, value={"state": state, "lt": lt})

    if isinstance(step, StepEphemerisSpkezp):
        pos, lt = _normalize_pos_lt(
            sp.spkezp(step.target, step.et, step.ref, step.abcorr, step.observer),
            label="spkezp",
        )
        context.state.ephemeris.requestedTargets.add(step.target)
        return StepOutput(op=step.op, value={"pos": pos, "lt": lt})

    if isinstance(step, StepEphemerisSpkezr):
        state, lt = _normalize_state_lt(
            sp.spkezr(step.target, step.et, step.ref, step.abcorr, step.observer),
            label="spkezr",
        )
        return StepOutput(op=step.op, value={"state": state, "lt": lt})

    if isinstance(step, StepEphemerisSpkgeo):
        state, lt = _normalize_state_lt(
            sp.spkgeo(step.target, step.et, step.ref, step.observer),
            label="spkgeo",
        )
        context.state.ephemeris.requestedTargets.add(step.target)
        return StepOutput(op=step.op, value={"state": state, "lt": lt})

    if isinstance(step, StepEphemerisSpkgps):
        pos, lt = _normalize_pos_lt(
            sp.spkgps(step.target, step.et, step.ref, step.observer),
            label="spkgps",
        )
        context.state.ephemeris.requestedTargets.add(step.target)
        return StepOutput(op=step.op, value={"pos": pos, "lt": lt})

    if isinstance(step, StepEphemerisSpkobj):
        ids_cell = _get_int_cell(
            context,
            step.idsCellId,
            max_cardinality=step.maxCardinality if step.maxCardinality is not None else 1024,
        )
        sp.spkobj(_resolve_path(context, step.spk), ids_cell)
        ids = [int(value) for value in ids_cell]
        return StepOutput(op=step.op, value={"ids": ids})

    if isinstance(step, StepEphemerisSpkopa):
        handle = int(sp.spkopa(_resolve_path(context, step.file)))
        _set_spk_handle(context, step.handleId, handle)
        return StepOutput(op=step.op, value={"handleId": step.handleId})

    if isinstance(step, StepEphemerisSpkopn):
        output_path = Path(_resolve_path(context, step.file))
        output_path.parent.mkdir(parents=True, exist_ok=True)
        handle = int(sp.spkopn(str(output_path), step.ifname, step.ncomch))
        _set_spk_handle(context, step.handleId, handle)
        return StepOutput(op=step.op, value={"handleId": step.handleId})

    if isinstance(step, StepEphemerisSpkpds):
        packed = _to_float_list(
            sp.spkpds(step.body, step.center, step.frame, step.type, step.first, step.last),
            expected_len=5,
            label="spkpds",
        )
        return StepOutput(op=step.op, value=packed)

    if isinstance(step, StepEphemerisSpkpos):
        pos, lt = _normalize_pos_lt(
            sp.spkpos(step.target, step.et, step.ref, step.abcorr, step.observer),
            label="spkpos",
        )
        return StepOutput(op=step.op, value={"pos": pos, "lt": lt})

    if isinstance(step, StepEphemerisSpksfs):
        try:
            out = _normalize_spksfs_result(sp.spksfs(step.body, step.et, 256))
            return StepOutput(op=step.op, value=out)
        except NotFoundError:
            return StepOutput(op=step.op, value={"found": False})

    if isinstance(step, StepEphemerisSpkssb):
        state = _to_float_list(sp.spkssb(step.target, step.et, step.ref), expected_len=6, label="spkssb")
        context.state.ephemeris.requestedTargets.add(step.target)
        return StepOutput(op=step.op, value=state)

    if isinstance(step, StepEphemerisSpkuds):
        body, center, frame, typenum, first, last, baddr, eaddr = sp.spkuds(step.descr)
        return StepOutput(
            op=step.op,
            value={
                "body": int(body),
                "center": int(center),
                "frame": int(frame),
                "type": int(typenum),
                "first": float(first),
                "last": float(last),
                "baddr": int(baddr),
                "eaddr": int(eaddr),
            },
        )

    if isinstance(step, StepEphemerisSpkw08):
        handle = _require_spk_handle(context, step.handleId)
        if len(step.states) == 0 or len(step.states) % 6 != 0:
            raise ValueError("spkw08.states must be a non-empty flat list whose length is a multiple of 6")

        state_records = [step.states[idx : idx + 6] for idx in range(0, len(step.states), 6)]
        sp.spkw08(
            handle,
            step.body,
            step.center,
            step.frame,
            step.first,
            step.last,
            step.segid,
            step.degree,
            len(state_records),
            state_records,
            step.epoch1,
            step.step,
        )
        return StepOutput(op=step.op, value=None)

    return None
